File: src/lib/converter/map-exporter/run-map-generate.ts
import { rmSync } from 'fs';
import path from 'path';

import { Config } from '@/lib/global-config';
import { normalizeMapSaveName } from '@/lib/map-save-name';
import { assertWowCascReady } from '@/lib/wow/wow-config-service';

import {
  defaultMapExportConfig,
  MapExportConfig,
  MapExporter,
} from './map-exporter';
import {
  autoChooseClampPercent,
  computeCreatureExportSteps,
  countUniqueUnitExportsFromManager,
  logMapGeneratePhase,
  pruneDepth,
} from './map-generate-utils';

const DEFAULT_PRUNE_DEPTH_SHELLS = 2;
const DEFAULT_PRUNE_DEPTH_INTERIORS = 3;

export interface MapGenerateConversionOptions {
  config: Config;
  mapExportConfig: MapExportConfig;
  mapSaveName: string;
  freshExport: boolean;
  includeBuildingInteriors?: boolean;
  autoClampPercent?: boolean;
  unitScale: number;
  onConvertStepsKnown?: (convertSteps: number) => void;
  onProgress?: (
    convertCompletedSteps: number,
    taskName: string,
    creatureProgress?: { completed: number; total: number },
  ) => void;
}

export interface MapGenerateConversionResult {
  outputDir: string;
  mapSaveName: string;
  convertSteps: number;
}

export async function runMapGenerateConversion(
  options: MapGenerateConversionOptions,
): Promise<MapGenerateConversionResult> {
  const {
    config,
    mapExportConfig,
    mapSaveName: rawSaveName,
    freshExport,
    includeBuildingInteriors = true,
    autoClampPercent = true,
    unitScale,
    onConvertStepsKnown,
    onProgress,
  } = options;

  const mapSaveName = normalizeMapSaveName(rawSaveName);
  const outputDir = path.join('maps', mapSaveName);

  await assertWowCascReady();

  if (freshExport) {
    try {
      rmSync(outputDir, { recursive: true, force: true });
      logMapGeneratePhase(`Removed existing map folder ${outputDir}`);
    } catch {
      // ignore
    }
  }

  const conversionConfig: Config = {
    ...config,
    isBulkExport: true,
    overrideModels: freshExport,
    mdx: true,
  };

  let convertCompleted = 0;
  const report = (
    taskName: string,
    creatureProgress?: { completed: number; total: number },
  ) => {
    onProgress?.(convertCompleted, taskName, creatureProgress);
  };

  const mapExporter = new MapExporter(conversionConfig, mapExportConfig);

  logMapGeneratePhase('Parsing map objects');
  report('Parsing map data');
  await mapExporter.parseObjects();
  pruneDepth(
    mapExporter,
    includeBuildingInteriors ? DEFAULT_PRUNE_DEPTH_INTERIORS : DEFAULT_PRUNE_DEPTH_SHELLS,
  );

  if (autoClampPercent) {
    autoChooseClampPercent(mapExporter, mapExportConfig, unitScale);
  }

  convertCompleted = 1;
  report('Parsed map data');

  const uniqueCreatureCount = mapExportConfig.creatures.enable
    ? countUniqueUnitExportsFromManager(mapExporter.wowObjectManager)
    : 0;
  const creatureExportSteps = computeCreatureExportSteps(uniqueCreatureCount);
  const convertSteps = 1 + 1 + creatureExportSteps + 1;
  onConvertStepsKnown?.(convertSteps);

  logMapGeneratePhase('Saving terrain and doodads');
  report('Saving terrain and doodads');
  await mapExporter.exportTerrainsDoodads(outputDir);
  convertCompleted = 2;
  report('Saved terrain and doodads');

  if (mapExportConfig.creatures.enable && creatureExportSteps > 0) {
    logMapGeneratePhase('Exporting creature models');
    report('Exporting creature models', { completed: 0, total: uniqueCreatureCount });
    await mapExporter.exportCreatures(outputDir, {
      onCreatureProgress: (completed, total) => {
        convertCompleted = 2 + completed;
        report('Exporting creature models', { completed, total });
      },
    });
    convertCompleted = 2 + creatureExportSteps;
    report('Exported creature models');
  }

  logMapGeneratePhase('Saving war3map files');
  report('Saving war3map files');
  mapExporter.saveWar3mapFiles(outputDir, mapSaveName);
  convertCompleted = convertSteps;
  report('Complete');

  return { outputDir: path.resolve(outputDir), mapSaveName, convertSteps };
}

export function buildMapExportConfig(params: {
  mapId: number;
  wowExportFolder: string;
  min: [number, number];
  max: [number, number];
  mapAngleDeg: number;
  clampLower: number;
  clampUpper: number;
  unitScale: number;
  creatures: MapExportConfig['creatures'];
}): MapExportConfig {
  return {
    ...defaultMapExportConfig,
    mapId: params.mapId,
    wowExportFolder: params.wowExportFolder,
    min: params.min,
    max: params.max,
    mapAngleDeg: params.mapAngleDeg,
    terrain: {
      clampPercent: {
        lower: params.clampLower,
        upper: params.clampUpper,
      },
    },
    unitScale: params.unitScale,
    creatures: params.creatures,
  };
}

File: golang/internal/converter/mapexporter/run_map_generate.go
package mapexporter

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	"github.com/pqhuy98/wow-converter/internal/config"
	"github.com/pqhuy98/wow-converter/internal/math"
	"github.com/pqhuy98/wow-converter/internal/wow/client"
)

// MapGenerateConversionOptions configures a server-side WC3 conversion run.
type MapGenerateConversionOptions struct {
	Config                   config.Config
	MapExportConfig          *MapExportConfig
	MapSaveName              string
	FreshExport              bool
	IncludeBuildingInteriors bool
	AutoClampPercent         bool
	UnitScale                float64
	WowClient                client.Client
	OnConvertStepsKnown      func(convertSteps int)
	OnProgress               func(convertCompleted int, taskName string, creatureProgress *CreatureProgress)
}

// CreatureProgress tracks creature export progress.
type CreatureProgress struct {
	Completed int
	Total     int
}

// MapGenerateConversionResult is the output of runMapGenerateConversion.
type MapGenerateConversionResult struct {
	OutputDir    string
	MapSaveName  string
	ConvertSteps int
}

// BuildMapExportConfig constructs MapExportConfig from API parameters.
func BuildMapExportConfig(params struct {
	MapID           int
	WowExportFolder string
	Min             math.Vector2
	Max             math.Vector2
	MapAngleDeg     float64
	ClampLower      float64
	ClampUpper      float64
	UnitScale       float64
	CreaturesEnable bool
	AllAreDoodads   bool
}) MapExportConfig {
	cfg := DefaultMapExportConfig()
	cfg.MapID = params.MapID
	cfg.WowExportFolder = params.WowExportFolder
	cfg.Min = params.Min
	cfg.Max = params.Max
	cfg.MapAngleDeg = params.MapAngleDeg
	cfg.Terrain.ClampPercent.Lower = params.ClampLower
	cfg.Terrain.ClampPercent.Upper = params.ClampUpper
	cfg.UnitScale = params.UnitScale
	cfg.Creatures.Enable = params.CreaturesEnable
	cfg.Creatures.AllAreDoodads = params.AllAreDoodads
	return cfg
}

// RunMapGenerateConversion executes parse, export, and save for a generated map.
func RunMapGenerateConversion(ctx context.Context, opts MapGenerateConversionOptions) (MapGenerateConversionResult, error) {
	autoClamp := opts.AutoClampPercent

	if opts.WowClient != nil {
		if err := opts.WowClient.WaitUntilReady(ctx); err != nil {
			return MapGenerateConversionResult{}, fmt.Errorf("WoW data not loaded: %w", err)
		}
	}

	mapSaveName := NormalizeMapSaveName(opts.MapSaveName)
	outputDir := filepath.Join("maps", mapSaveName)
	if opts.FreshExport {
		_ = os.RemoveAll(outputDir)
		LogMapGeneratePhase("Removed existing map folder " + outputDir)
	}

	conversionCfg := opts.Config
	conversionCfg.IsBulkExport = true
	conversionCfg.OverrideModels = opts.FreshExport
	conversionCfg.MDX = true

	convertCompleted := 0
	report := func(taskName string, creature *CreatureProgress) {
		if opts.OnProgress != nil {
			opts.OnProgress(convertCompleted, taskName, creature)
		}
	}

	exporter := NewMapExporter(conversionCfg, opts.MapExportConfig, opts.WowClient)
	LogMapGeneratePhase("Parsing map objects")
	report("Parsing map data", nil)
	if err := exporter.ParseObjects(nil); err != nil {
		return MapGenerateConversionResult{}, err
	}
	pruneDepth := 2
	if opts.IncludeBuildingInteriors {
		pruneDepth = 3
	}
	PruneDepth(exporter, pruneDepth)
	if autoClamp {
		AutoChooseClampPercent(exporter, opts.UnitScale)
	}
	convertCompleted = 1
	report("Parsed map data", nil)

	uniqueCreatureCount := 0
	if opts.MapExportConfig != nil && opts.MapExportConfig.Creatures.Enable {
		uniqueCreatureCount = CountUniqueUnitExportsFromManager(exporter.WowObjectManager)
	}
	creatureExportSteps := ComputeCreatureExportSteps(uniqueCreatureCount)
	convertSteps := 1 + 1 + creatureExportSteps + 1
	if opts.OnConvertStepsKnown != nil {
		opts.OnConvertStepsKnown(convertSteps)
	}

	LogMapGeneratePhase("Saving terrain and doodads")
	report("Saving terrain and doodads", nil)
	if err := exporter.ExportTerrainsDoodads(outputDir); err != nil {
		return MapGenerateConversionResult{}, err
	}
	convertCompleted = 2
	report("Saved terrain and doodads", nil)

	if opts.MapExportConfig != nil && opts.MapExportConfig.Creatures.Enable && creatureExportSteps > 0 {
		LogMapGeneratePhase("Exporting creature models")
		report("Exporting creature models", &CreatureProgress{Completed: 0, Total: uniqueCreatureCount})
		if err := exporter.ExportCreatures(outputDir, func(completed, total int) {
			convertCompleted = 2 + completed
			report("Exporting creature models", &CreatureProgress{Completed: completed, Total: total})
		}); err != nil {
			return MapGenerateConversionResult{}, err
		}
		convertCompleted = 2 + creatureExportSteps
		report("Exported creature models", nil)
	}

	LogMapGeneratePhase("Saving war3map files")
	report("Saving war3map files", nil)
	if err := exporter.SaveWar3mapFiles(outputDir, mapSaveName); err != nil {
		return MapGenerateConversionResult{}, err
	}
	convertCompleted = convertSteps
	report("Complete", nil)

	absOut, _ := filepath.Abs(outputDir)
	LogMapGeneratePhase("Complete: " + mapSaveName + " → " + absOut)
	return MapGenerateConversionResult{
		OutputDir:    absOut,
		MapSaveName:  mapSaveName,
		ConvertSteps: convertSteps,
	}, nil
}
